### PR TITLE
Allow assiging/unassigning classes to schools

### LINKED ISSUE #5412: [ocis] libregraph: Allow assigning/unassigning classes to schools
The `/education/schools/{school-id}/classes` and `/education/schools/{school-id}/classes/{class-id}` related endpoints still need to be implemented.

## Definition of ready
[ ] everybody needs to understand the value written in the user story
[ ] acceptance criteria has to be defined
[ ] all dependencies of the user story need to be identified
[ ] feature should be seen from an end user perspective
[ ] user story has to be estimated
[ ] story points need to be less then 20

## Definition of done
- Functional requirements
[ ] functionality described in the user story works
[ ] acceptance criteria are fulfilled
- Quality
[ ] code review happened
[ ] CI is green
[ ] critical code received unit tests by the developer
[ ] automated tests passed (if automated tests are not available, this test needs to be created and passed
- Non-functional requirements
[ ] no sonar cloud issues

### PR DESCRIPTION
**Pull request by ainmosni**
This PR does the following:

Adds handlers for `/education/schools/{schoolID}/classes/{$ref,classID,}` to allow the requesting and manipulation of classes.

This fixes #5412

### CODE REVIEW

**Comment by ownclouders** [comment]
:boom: Acceptance test [<strong>settingsUITests</strong>](https://drone.owncloud.com/owncloud/ocis/18558/62/1) failed. Further test are cancelled...

File: services/graph/pkg/identity/ldap_education_school.go [hunk 1]
```diff
@@ -479,6 +454,154 @@ func (i *LDAP) RemoveUserFromEducationSchool(ctx context.Context, schoolNumberOr
...
+	logger.Debug().Str("backend", "ldap").Msg("GetEducationSchoolClasses")
+
+	entries, err := i.getEducationSchoolEntries(
+		schoolNumberOrID, i.groupFilter, i.educationConfig.classObjectClass, i.groupBaseDN, i.groupScope, i.getEducationClassAttrTypes(false), logger, // TODO: Add attributes
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	users := make([]*libregraph.EducationClass, 0, len(entries))
```
Comment (rhafer) [inline services/graph/pkg/identity/ldap_education_school.go:469]:
```suggestion
	classes := make([]*libregraph.EducationClass, 0, len(entries))
```

File: services/graph/pkg/identity/ldap_education_school.go [hunk 2]
```diff
@@ -479,6 +454,154 @@ func (i *LDAP) RemoveUserFromEducationSchool(ctx context.Context, schoolNumberOr
...
+
+	for _, e := range entries {
+		u := i.createEducationClassModelFromLDAP(e)
+		// Skip invalid LDAP classes
+		if u == nil {
+			continue
+		}
+		users = append(users, u)
```
Comment (rhafer) [inline services/graph/pkg/identity/ldap_education_school.go:477]:
```suggestion
		class := i.createEducationClassModelFromLDAP(e)
		// Skip invalid LDAP classes
		if class == nil {
			continue
		}
		classes = append(classes, class)
```

File: services/graph/pkg/identity/ldap_education_school.go [hunk 3]
```diff
@@ -479,6 +454,154 @@ func (i *LDAP) RemoveUserFromEducationSchool(ctx context.Context, schoolNumberOr
...
+	}
+	return users, nil
```
Comment (rhafer) [inline services/graph/pkg/identity/ldap_education_school.go:479]:
```suggestion
	return classes, nil
```

File: services/graph/pkg/identity/ldap_education_school.go [hunk 4]
```diff
@@ -479,6 +454,154 @@ func (i *LDAP) RemoveUserFromEducationSchool(ctx context.Context, schoolNumberOr
 	return nil
 }
 
+// GetEducationSchoolClasses implements the EducationBackend interface for the LDAP backend.
+func (i *LDAP) GetEducationSchoolClasses(ctx context.Context, schoolNumberOrID string) ([]*libregraph.EducationClass, error) {
+	logger := i.logger.SubloggerWithRequestID(ctx)
+	logger.Debug().Str("backend", "ldap").Msg("GetEducationSchoolClasses")
+
+	entries, err := i.getEducationSchoolEntries(
+		schoolNumberOrID, i.groupFilter, i.educationConfig.classObjectClass, i.groupBaseDN, i.groupScope, i.getEducationClassAttrTypes(false), logger, // TODO: Add attributes
```
Comment (rhafer) [inline services/graph/pkg/identity/ldap_education_school.go:463]: What is the `TODO` referring to?

File: services/graph/pkg/identity/ldap_education_school.go [hunk 5]
```diff
@@ -479,6 +454,154 @@ func (i *LDAP) RemoveUserFromEducationSchool(ctx context.Context, schoolNumberOr
...
+		return ErrNotFound
+	}
+
+	schoolID := schoolEntry.GetEqualFoldAttributeValue(i.educationConfig.schoolAttributeMap.id)
+	user, err := i.getEducationClassByID(memberID, false)
+	if err != nil {
+		i.logger.Warn().Str("userid", memberID).Msg("Class does not exist")
+		return err
+	}
+	currentSchools := user.GetEqualFoldAttributeValues(i.educationConfig.memberOfSchoolAttribute)
```
Comment (rhafer) [inline services/graph/pkg/identity/ldap_education_school.go:591]: s/user/class/

**Review by rhafer — CHANGES_REQUESTED** [review]

File: services/graph/pkg/identity/ldap_education_school.go (hunk 4 above)
Comment (ainmosni) [reply services/graph/pkg/identity/ldap_education_school.go]: That was for myself, I forgot to remove it.

File: services/graph/pkg/identity/ldap_education_school.go [hunk 6]
```diff
@@ -479,6 +454,154 @@ func (i *LDAP) RemoveUserFromEducationSchool(ctx context.Context, schoolNumberOr
...
+	if err != nil {
+		return err
+	}
+
+	if schoolEntry == nil {
+		return ErrNotFound
+	}
+
+	schoolID := schoolEntry.GetEqualFoldAttributeValue(i.educationConfig.schoolAttributeMap.id)
+	class, err := i.getEducationClassByID(memberID, false)
```
Comment (rhafer) [inline services/graph/pkg/identity/ldap_education_school.go:586]: It seems that `getEducationClassByID` does not request the `ocMemberOfSchool` Attribute. So the for loop below does actually never do anything. (The bug is probably in `getEducationClassAttrTypes`)

**Review by rhafer — CHANGES_REQUESTED** [review]

**Review by rhafer — APPROVED** [review]

**Pull request merged**